browser/static/js/main_zstack.js

```diff
@@ -751,14 +697,14 @@ function upload_file(cb) {
   $.ajax({
     type: 'POST',
     url: `${document.location.origin}/upload_file/${project_id}`,
-    success: cb,
     async: true
-  });
+  }).done(cb);
 }
 
 function changeZoom(dzoom) {
-  state.changeZoom(dzoom, state.canvasPosX, state.canvasPosY);
-  updateMousePos(state.rawX, state.rawY);
+  zoom = new Zoom(canvas, dzoom);
+  updateMousePos(canvas.rawX, canvas.rawY);
```

Comment:
Hmmm I'm struggling with this one... updateMousePos interacts with multiple global objects (both canvas and brush), and is used both by moving the mouse and by Zooming.

Given that undo/redoing clears the mode (& the selected labels), I'm not sure if it's possible to do an action immediately after undoing without moving the mouse and updating the mouse position. The brush also shows the old location until its updated, so it is still clear which label is selected, even if the cursor is not lined up with it.

I'm going to not address this issue because it would require too much restructuring of the global objects and their interactions.

Reply:
I took another look, and I think there may be a reasonable solution by updating the mouse position after any undo/redo, not just Zooms. I got tripped up trying to record the mouse history in the Zoom, and I think this is something that should not be tracked by undo/redo, just made consistent again. I think with this change the behavior works as expected.